### DISCUSSION TITLE
Enable/disable copilot suggestions

### DISCUSSION THREAD

**Discussion by yngwi** [saghen/blink.cmp#226] [Q&A] [answered]
Hi, I’m trying to integrate with copilot.lua. I have seen a snippet in this [Pull Request](https://github.com/Saghen/blink.cmp/pull/31) that I would like to integrate. When using `require("blink.cmp").on_open` then I get an error informing me that `windows` is a nil value. So i tried to read the code and use the internal functionality for now:

```lua
local autocomplete = require("blink.cmp.windows.autocomplete")
autocomplete.listen_on_open(function()
	require("copilot.suggestion").dismiss()
	vim.api.nvim_buf_set_var(0, "copilot_suggestion_hidden", true)
end)

autocomplete.listen_on_close(function()
	vim.api.nvim_buf_set_var(0, "copilot_suggestion_hidden", false)
end)
```

However, doing this I noticed that the callback seems to be triggered each time the autocomplete *could* be opened and not each time it is actually opened because I have `auto_show = false`. This means that the code gets executed whenever I press a key in insert mode as that would generally trigger the autocomplete. Is this a misunderstanding on my part on how to correctly use this or a bug?

**Comment by saghen** [comment]
For `window` being nil, you'll want to make sure you call `.setup` before hooking in any listeners. WRT `auto_show = false` still triggering the event, that sounds like a bug that #204 will fix

**Reply by yngwi** [reply]
Thank you for the prompt reply. W/r to the window problem, I call the hooks in the lazy config function directly after the setup. I’m not too knowledgable, so this might be wrong.

**Comment by collindutter** [comment, marked as answer]
This is how I implemented it, seems to be working well.

```lua
    vim.api.nvim_create_autocmd('User', {
      pattern = 'BlinkCmpCompletionMenuOpen',
      callback = function()
        vim.b.copilot_suggestion_hidden = true
      end,
    })

    vim.api.nvim_create_autocmd('User', {
      pattern = 'BlinkCmpCompletionMenuClose',
      callback = function()
        vim.b.copilot_suggestion_hidden = false
      end,
    })

```

Thanks for the amazing plugin @Saghen.

**Reply by yngwi** [reply]
Currently i do this:

```lua
local cmp = require("blink.cmp.completion.list")

-- Handler for show event
cmp.show_emitter:on(function()
	require("copilot.suggestion").dismiss()
	vim.api.nvim_buf_set_var(0, "copilot_suggestion_hidden", true)
end)

-- Handler for hide event
cmp.hide_emitter:on(function()
	vim.api.nvim_buf_set_var(0, "copilot_suggestion_hidden", false)
end)
```

I think I like yours better :D
